include/private/hashmap.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Hash map support
+ *
+ * Copyright 2020 Philip Rebohle for Valve Corporation
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#ifndef __VKD3D_HASHMAP_H
+#define __VKD3D_HASHMAP_H
+
+#include <stddef.h>
+
+#include "memory.h"
+
+enum hash_map_entry_flag
+{
+    HASH_MAP_ENTRY_OCCUPIED = (1 << 0),
+};
+
+struct hash_map_entry
+{
+    uint32_t hash_value;
+    uint32_t flags;
+};
+
+typedef uint32_t (*pfn_hash_func)(const void* key);
+typedef bool (*pfn_hash_compare_func)(const void *key, const struct hash_map_entry *entry);
+
+/* Open-addressing hash table */
+struct hash_map
+{
+    pfn_hash_func hash_func;
+    pfn_hash_compare_func compare_func;
+    void *entries;
+    size_t entry_size;
+    uint32_t entry_count;
+    uint32_t used_count;
+};
+
+static inline struct hash_map_entry *hash_map_get_entry(struct hash_map *hash_map, uint32_t entry_idx)
+{
+    return hash_map->entries + hash_map->entry_size * entry_idx;
+}
+
+static inline uint32_t hash_map_get_entry_idx(struct hash_map *hash_map, uint32_t hash_value)
+{
+    return hash_value % hash_map->entry_count;
+}
+
+static inline uint32_t hash_map_next_size(uint32_t old_size)
+{
+    /* This yields a sequence of primes and numbers with two
+     * relatively large prime factors for any reasonable hash
+     * table size */
+    return old_size ? (old_size * 2 + 5) : 37;
+}
+
+static inline bool hash_map_grow(struct hash_map *hash_map)
+{
+    uint32_t i, old_count, new_count;
+    void *new_entries, *old_entries;
+
+    old_count = hash_map->entry_count;
+    old_entries = hash_map->entries;
+
+    new_count = hash_map_next_size(hash_map->entry_count);
+
+    if (!(new_entries = vkd3d_calloc(new_count, hash_map->entry_size)))
+        return false;
+
+    hash_map->entry_count = new_count;
+    hash_map->entries = new_entries;
+
+    for (i = 0; i < old_count; i++)
+    {
+        /* Relocate existing entries one by one */
+        struct hash_map_entry *old_entry = old_entries + i * hash_map->entry_size;
+
+        if (old_entry->flags & HASH_MAP_ENTRY_OCCUPIED)
+        {
+            memcpy(hash_map_get_entry(hash_map, hash_map_get_entry_idx(hash_map, old_entry->hash_value)),
```

Comment:
Does this work? Can't there be collisions?

Reply:
Good point, will fix.